### ISSUE TITLE
Interface issues/suggestions

### ISSUE THREAD

**Issue by vertigo220** [mifi/lossless-cut#1859]
### I have a lot of issues to go through, so in order to make it easier for me to help you, I ask that you please try these things first

- [X] Check out all the menus and keyboard shortcuts. Maybe the feature already exists?
- [X] Try with the [newest version from GitHub](https://github.com/mifi/lossless-cut/releases/latest)
- [X] Read the [documentation](https://github.com/mifi/lossless-cut) and [FAQ, Known issues, Troubleshooting](https://github.com/mifi/lossless-cut/blob/master/issues.md) to see if the feature already exists
- [X] See if someone already suggested this under [Issues](https://github.com/mifi/lossless-cut/issues) or [Discussions](https://github.com/mifi/lossless-cut/discussions)

### Description

I've found a handful of things regarding the interface that I feel could be improved to make using LC a smoother experience. Instead of making a separate issue for each, I'll just lay them all out here. If separate issues are preferred, let me know.

- There's no option to use the mouse scroll wheel to go frame-by-frame. Ideally, it should allow both the use of a modifier key to perform the alternate behavior from the default as well as provide a button to toggle the default, e.g. scroll without a modifier to go forward/back by keyframe and with a modifier to do so by frame, and click a button, like the one to switch the keep/delete selected segments mode, to flip that behavior.
- When going into certain settings, e.g. the "Change value" ones under the "Keyboard, mouse and input" section, once the setting is adjusted and the user clicks "Done" it should go back to the settings where they were.
- When setting keyboard shortcuts, if attempting to add a key already in use to another function, it should notify the user of this, and what function it's currently assigned to, instead of simply doing nothing.
- When showing the "Export options" window before exporting and opening the track selection screen, this screen is both a bit confusing and inconsistent with others throughout LC in that it requires closing it via the x to apply the changes (vs clicking a "Done" or similar button) and resulted in my clicking the check button ("Keep all tracks") thinking that was the confirm button. Also, the second icon/button at the top-right (between the file info and file metadata ones) doesn't have a tooltip. Finally, and most importantly, once the selections are made and this screen is closed, instead of taking the user back to the export options window like it should, it goes all the way back to the main interface.
- Very minor, but I prefer buttons for dialog windows to be in/near those windows, and find the "Export" button being all the way at the bottom-right to be a bit awkward. Perhaps because of the size of my monitor and the fact I have a lot going on on my computer and have stuff on my desk that partially blocks that corner of the screen, but it took me a few seconds to figure out what was going on as I was expecting the button to be at the bottom of the export options dialog, and especially since I had just had to close the track selection window to apply those changes, I thought maybe this was similar and had to be closed to start it. Certainly something that despite the initial confusion would be fine once used to it, but worth mentioning.

Aside from these minor issues (the first one is the only one I'd consider a must-have, whereas the others are just minor inconveniences), this appears to be a very well-designed and intuitive program. I've used multiple editing programs before, and I just tried a couple other popular ones and couldn't even figure out how to do basic timeline trimming, whereas this was extremely easy to just jump right into. Excellent job, and thank you.

**Comment by mifi** [comment]
thanks for your suggestions

> * There's no option to use the mouse scroll wheel to go frame-by-frame. Ideally, it should allow both the use of a modifier key to perform the alternate behavior from the default as well as provide a button to toggle the default, e.g. scroll without a modifier to go forward/back by keyframe and with a modifier to do so by frame, and click a button, like the one to switch the keep/delete selected segments mode, to flip that behavior.

it works for me on macOs. are you saying that when you use the mouse wheel, nothing happens? if so, which OS are you on? have you tried to adjust the trackpad/wheel speed in settings?

> * When going into certain settings, e.g. the "Change value" ones under the "Keyboard, mouse and input" section, once the setting is adjusted and the user clicks "Done" it should go back to the settings where they were.

probably won't prioritize that because it's not something I'd imagine people use that often

> * When setting keyboard shortcuts, if attempting to add a key already in use to another function, it should notify the user of this, and what function it's currently assigned to, instead of simply doing nothing.

this has been fixed (in an unreleased version)

> * When showing the "Export options" window before exporting and opening the track selection screen, this screen is both a bit confusing and inconsistent with others throughout LC in that it requires closing it via the x to apply the changes (vs clicking a "Done" or similar button) and resulted in my clicking the check button ("Keep all tracks") thinking that was the confirm button.

not sure how to solve this. the X button is common in many dialogs in loslesscut

> * Also, the second icon/button at the top-right (between the file info and file metadata ones) doesn't have a tooltip. Finally, and most importantly, once the selections are made and this screen is closed, instead of taking the user back to the export options window like it should, it goes all the way back to the main interface.

fixed

> * Very minor, but I prefer buttons for dialog windows to be in/near those windows, and find the "Export" button being all the way at the bottom-right to be a bit awkward. Perhaps because of the size of my monitor and the fact I have a lot going on on my computer and have stuff on my desk that partially blocks that corner of the screen, but it took me a few seconds to figure out what was going on as I was expecting the button to be at the bottom of the export options dialog, and especially since I had just had to close the track selection window to apply those changes, I thought maybe this was similar and had to be closed to start it. Certainly something that despite the initial confusion would be fine once used to it, but worth mentioning.

I think because the export button is at the very lower right bottom before opening the export dialog, it also makes sense that it is in the same location once the button is clicked and the export dialog is shown

**Comment by vertigo220** [comment]
> thanks for your suggestions
> 
> > * There's no option to use the mouse scroll wheel to go frame-by-frame. Ideally, it should allow both the use of a modifier key to perform the alternate behavior from the default as well as provide a button to toggle the default, e.g. scroll without a modifier to go forward/back by keyframe and with a modifier to do so by frame, and click a button, like the one to switch the keep/delete selected segments mode, to flip that behavior.
> 
> it works for me on macOs. are you saying that when you use the mouse wheel, nothing happens? if so, which OS are you on? have you tried to adjust the trackpad/wheel speed in settings?

It moves forward and backward, but by keyframe. I'm saying I'd like to be able to move forward and backward frame-by-frame with the mouse wheel, and it would be ideal to be able to do this in two ways: by holding a user-customizable modifier key while scrolling, and to have a toggle button in the UI to switch between keyframe and frame.

> > * When going into certain settings, e.g. the "Change value" ones under the "Keyboard, mouse and input" section, once the setting is adjusted and the user clicks "Done" it should go back to the settings where they were.
> 
> probably won't prioritize that because it's not something I'd imagine people use that often

Agree not a priority, but possibly some day.

> > * When setting keyboard shortcuts, if attempting to add a key already in use to another function, it should notify the user of this, and what function it's currently assigned to, instead of simply doing nothing.
> 
> this has been fixed (in an unreleased version)

Related to this, and maybe the fix has already done this, it shouldn't just show what it's currently used for but allow replacing it as well.

> > * When showing the "Export options" window before exporting and opening the track selection screen, this screen is both a bit confusing and inconsistent with others throughout LC in that it requires closing it via the x to apply the changes (vs clicking a "Done" or similar button) and resulted in my clicking the check button ("Keep all tracks") thinking that was the confirm button.
> 
> not sure how to solve this. the X button is common in many dialogs in loslesscut

Could just add an OK button that does the same as closing the window, so either works. Though closing a window is typically thought to cancel any changes, as that's the standard across most apps, so I'd still think people may do that in an attempt to cancel the changes they've made.

> > * Very minor, but I prefer buttons for dialog windows to be in/near those windows, and find the "Export" button being all the way at the bottom-right to be a bit awkward. Perhaps because of the size of my monitor and the fact I have a lot going on on my computer and have stuff on my desk that partially blocks that corner of the screen, but it took me a few seconds to figure out what was going on as I was expecting the button to be at the bottom of the export options dialog, and especially since I had just had to close the track selection window to apply those changes, I thought maybe this was similar and had to be closed to start it. Certainly something that despite the initial confusion would be fine once used to it, but worth mentioning.
> 
> I think because the export button is at the very lower right bottom before opening the export dialog, it also makes sense that it is in the same location once the button is clicked and the export dialog is shown

The problem is, once the dialog is shown, the user would expect, as is the standard, that any buttons associated with it would be *in* it. It's strange to have a button in the main interface still be used with a dialog present like that. That said, it's a minor issue and I've provided my thoughts and reasoning, so I'll leave it at that. Perhaps others can weigh in one way or the other.

**Comment by mifi** [comment]
> It moves forward and backward, but by keyframe. I'm saying I'd like to be able to move forward and backward frame-by-frame with the mouse wheel, and it would be ideal to be able to do this in two ways: by holding a user-customizable modifier key while scrolling, and to have a toggle button in the UI to switch between keyframe and frame.

hmm, for me it jumps only a small distance when i trackpad. maybe it's dependent on mouse hardware,driver and OS settings? if i set Timeline trackpad/wheel sensitivity to 0.01 i can even seek with a sub-frame accuracy

> Related to this, and maybe the fix has already done this, it shouldn't just show what it's currently used for but allow replacing it as well.

ideally yes, but not high priority

> Though closing a window is typically thought to cancel any changes, as that's the standard across most apps, so I'd still think people may do that in an attempt to cancel the changes they've made.

not sure I agree. I feel that these days when closing windows, settings dialogs or browser tabs, most apps will remember whichever changes you did. And if closing means you lose something, there will almost always be a popup confirmation dialog asking "are you sure you want to lose your unsaved data?".

Because almost all other dialogs in losslesscut behave like this (close means done, *not* lose all unsaved data), I think it makes sense for the tracks panel also to behave the same for consistency. if we add an OK button to the tracks panel, shouldn't we also add an OK button to close the export dialog (without exporting) and an OK button in the settings page

> The problem is, once the dialog is shown, the user would expect, as is the standard, that any buttons associated with it would be _in_ it. It's strange to have a button in the main interface still be used with a dialog present like that. That said, it's a minor issue and I've provided my thoughts and reasoning, so I'll leave it at that. Perhaps others can weigh in one way or the other.

i agree it's a bit weird, but i still think it's ok considering the fact that the user already clicked that location, and is likely to click again without moving their mouse to complete the export

**Comment by vertigo220** [comment]
> > It moves forward and backward, but by keyframe. I'm saying I'd like to be able to move forward and backward frame-by-frame with the mouse wheel, and it would be ideal to be able to do this in two ways: by holding a user-customizable modifier key while scrolling, and to have a toggle button in the UI to switch between keyframe and frame.
> 
> hmm, for me it jumps only a small distance when i trackpad. maybe it's dependent on mouse hardware,driver and OS settings? if i set Timeline trackpad/wheel sensitivity to 0.01 i can even seek with a sub-frame accuracy

A few things here. First, I didn't even think to try that; I guess it just seemed like it did something else. Perhaps it would be helpful to put an explanation in small text under that (and other?) option(s). I tried adjusting it, and it did help, but even at 0.01, it still jumps ~5 frames at a time. So maybe the solution is to allow even finer gradations, but even then if it's not specifically advancing by frame, and is rather using some formula to determine how far it moves, and if that formula's result happens to be <=1 then it moves one frame, but if it happens to be >1 (or 1.5 or whatever) then it moves by two frames, that's still not really ideal, because it may from time to time move by more than one frame when I'm trying to go a frame at a time. So what it really needs is a specific function to go forward/back by one frame, regardless of anything else like sensitivity. And a way to quickly and easily toggle between modes, because sometimes I may want to advance by larger amounts, and I wouldn't want to have to go into settings and modify the sensitivity every time. Ideally, I could set it to go frame-by-frame with the mouse wheel, go by keyframe by holding shift while scrolling, by e.g. 200 frame by using ctrl, and by e.g. 2 seconds by using alt, and to be able to click a button on the UI to toggle the non-modified function with one of the modified functions, so I could click that then be able to scroll by keyframe without having to hold the shift key down.

> > Though closing a window is typically thought to cancel any changes, as that's the standard across most apps, so I'd still think people may do that in an attempt to cancel the changes they've made.
> 
> not sure I agree. I feel that these days when closing windows, settings dialogs or browser tabs, most apps will remember whichever changes you did. And if closing means you lose something, there will almost always be a popup confirmation dialog asking "are you sure you want to lose your unsaved data?".
> 
> Because almost all other dialogs in losslesscut behave like this (close means done, _not_ lose all unsaved data), I think it makes sense for the tracks panel also to behave the same for consistency. if we add an OK button to the tracks panel, shouldn't we also add an OK button to close the export dialog (without exporting) and an OK button in the settings page
> 
> > The problem is, once the dialog is shown, the user would expect, as is the standard, that any buttons associated with it would be _in_ it. It's strange to have a button in the main interface still be used with a dialog present like that. That said, it's a minor issue and I've provided my thoughts and reasoning, so I'll leave it at that. Perhaps others can weigh in one way or the other.
> 
> i agree it's a bit weird, but i still think it's ok considering the fact that the user already clicked that location, and is likely to click again without moving their mouse to complete the export

I get your point as well. All I'm saying is it threw me. I use a computer daily, home and work, with numerous various programs ranging from very simple to very complex, intuitive and not. When I first used LC, these aspects were different from what I'm used to in most programs and I personally found them awkward and unintuitive. As I said, I actually clicked the button for a completely different action, despite being in a strange place for an apply button, because it was the only thing I could see that made any sense at all. And yes, other areas are like that, but I also found those a bit strange and confusing at first. I just think it's helpful to give first impressions to get an idea of what something is like to a new user, and those were mine. You don't agree, and others may not as well, and that's fine, I just wanted to offer that feedback.

**Comment by mifi** [comment]
> I tried adjusting it, and it did help, but even at 0.01, it still jumps ~5 frames at a time

I will try to make it possible to set even more precise low values. I guess I never noticed because I only tested on macos with a trackpad which has very fine resolution on the "wheel" event.

> So maybe the solution is to allow even finer gradations, but even then if it's not specifically advancing by frame, and is rather using some formula to determine how far it moves, and if that formula's result happens to be <=1 then it moves one frame, but if it happens to be >1 (or 1.5 or whatever) then it moves by two frames, that's still not really ideal, because it may from time to time move by more than one frame when I'm trying to go a frame at a time.

it's actually a bit complex how wheel events are being transformed into scalar values, there's this code that handles it, which was added by request from someone, and which I don't understand 100%:

https://github.com/mifi/lossless-cut/blob/master/src/hooks/normalizeWheel.js

...so I don't really know how to transfer a "wheel" event into a fixed unit of say `1` frame. I think also different mouse wheels have different physical step sizes (e.g. the physical length that the wheel `snaps` into), so I'm not sure if it's even possible to make an accurate transformation like this.

> So what it really needs is a specific function to go forward/back by one frame, regardless of anything else like sensitivity.

There is already such a function, it's in keyboard shortcuts called "Step backward 1 frame" and "Step forward 1 frame", however I don't know how I would bind these to "wheel" events because I wouldn't know how to detect a "wheel" event (as described above).

I appreciate feedback. I will create a new issue where I will gather the suggestions from you as well as others about customising mouse events.
